### ISSUE TITLE
Get notification display datetime when app terminated

### ISSUE THREAD

**Issue by d-ukhanov** [MaikuB/flutter_local_notifications#2483]
I want to get the datetime when the notification was displayed to save it in the database. But how to do it when app terminated?
Is there any way to do this?

**Comment by Levi-Lesches** [comment]
Can you explain a bit more what you're asking? If you're trying to determine when a notification is displayed, can you just do that alongside the code that shows the notification? 

```dart
plugin.show(...);
final timestamp = DateTime.now();
```

If you're scheduling a notification in the future, just use the time you're scheduling it for

```dart
final timestamp = DateTime.now().add(Duration(days: 1));
plugin.zonedSchedule(..., TZDateTime.from(timestamp));
```

**Comment by d-ukhanov** [comment]
I need this if there are scheduled reminders for every day at some time. After receiving a notification and saving the data about it in the database, in the app there will be a note that the notification was, for example, "today at 10:00" or "on December 2 at 10:00".

**Comment by d-ukhanov** [comment]
As I understand it, the process of scheduling notifications once a day and other such does not allow for this. To schedule a daily notification you need to specify, for example, tomorrow's date and specify hours and minutes in it. And use DateTimeComponents.time. This way notifications will be triggered every day at the specified time. 
But it is not clear how to know on what day and what exact time each specific notification was displayed. Plus it is not always possible to display notifications in the same minute as it is scheduled, for many reasons because of OS, etc.

**Comment by d-ukhanov** [comment]
I've tried other options and libs, but haven't found a solution to this problem yet. For example, it could be that with the notification data comes the display time in onDidReceiveBackgroundNotificationResponse, but if the app is terminated, it will be null or equal to the time when the application goes to foreground or background, because onDidReceiveBackgroundNotificationResponse is triggered at that moment, if I understood correctly.

**Comment by d-ukhanov** [comment]
> "Like I said, I think you can get the date by just assuming the plugin worked as intended. So if you said "daily", you can display on your UI "today" or "yesterday" based on the current time of day."

I'm not sure I get it. For example, there is a scheduled notification that triggers every day. Today is December 4, and the user has not launched the app and ignored notifications since December 1 including. In other words, the application would like to display missed notifications from December 1-3, which were shown by the OS to the user. I.e. we plan to have something like a history of ignored notifications in the application.

> Let's say the user does not tap, and you do not use exact alarms. Maybe the device is off at the scheduled time. So we schedule for 5pm, and we get 8pm instead. The app is then launched at 9pm, and you want to know when the notification that was supposed to go off at 5pm actually went off, and you want some function to say "9pm". Is that what you're asking about?

So the notification itself only reached the user in 8pm? Then I would like 8pm in the record of this notification in the app.

By the way, while I was thinking about this, I realized there are still problems, unclear how to solve. For example, onDidReceiveBackgroundNotificationResponse and onDidReceiveNotificationResponse, as I understood, are triggered only if the user performs some actions, but how to get a callback when the notification has just been displayed? 
I'm not exactly sure, but doesn't the native code allow to make such things possible?
And maybe on Flutter it's possible, for example, to delay the callback for missed or ignored notifications until the app is launched?

As I understand it, that's not what this plugin currently offers, according by “do not support the concept of past notifications”. But maybe someone had a similar experience, would be curious to hear about it.

**Comment by Levi-Lesches** [comment]
> In other words, the application would like to display missed notifications

> we plan to have something like a history of ignored notifications in the application.

> but how to get a callback when the notification has just been displayed?

Seems like we're on the same page now 😅 

You're looking for a way to not only get when a notification was _supposed_ to be shown, but also when it was _actually_ shown. Or when it was "ignored" (by the way, you don't really define "ignored" -- left there for 1 second? 1 minute? 1 hour?)

What you're looking for is definitely not supported on Android. I don't know about iOS so I'll keep talking about Android: 

- the only class that can get alerted when a notification is posted is [`NotificationListenerService`](https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService#summary):
> To extend this class, you must declare the service in your manifest file with the [android.Manifest.permission#BIND_NOTIFICATION_LISTENER_SERVICE](https://developer.android.com/reference/kotlin/android/Manifest.permission.html#BIND_NOTIFICATION_LISTENER_SERVICE:kotlin.String)
- That link shows that only system-level apps -- apps written by the manufacturer of the device or Android itself -- can request this permission

So basically, what you're asking for is too sensitive to be sharing with random apps and is not allowed. It's not just that the _plugin_ doesn't support past notifications, the devices themselves do not. 

You're better off trying the "blind" way, of just showing the time you _intend_ to show the notifications and ignoring when they actually happened. If you want to be more exact, use the `exactAllowWhileIdle` option like I mentioned, but that's the best you can do.
